[Local GC] Enable building the GC as a shared library

swgillespie opened the pull request:
Unfortunately this is quite a large PR with a lot of small, unrelated changes (I'm sorry!) but I'll go through it in detail with comments. I'm sending this PR out early so that the CI can run on it and to solicit early reviews from anyone interested 😄 

This (currently WIP) PR enables the GC to be built as a shared library that is loaded through `LoadLibrary/dlopen` at startup. At this point all but 2 tests are passing on Windows when loading a GC dynamically and I'm in the process of running tests on OSX/Linux. Most of the work in this PR involved stubbing out or providing implementations for functionality that the GC previously received from the EE or the PAL - I'll point these out in comments.

swgillespie commented on CMakeLists.txt:
```diff
@@ -575,6 +575,13 @@ if(WIN32)
+#  - all clr specific compile definitions should be included in this file
+#  - all clr specific feature variable should also be added in this file
+#----------------------------------
+include(clrdefinitions.cmake)
```
The idea here is to move the inclusion of `clrdefinitions.cmake` up here to create an area of the CMake file where these definitions are active, but we have not yet included the PAL. This space (with CLR feature defs, but no PAL defs) is used to build the standalone GC.

There aren't any subdirectories being added between the old and new positions of `clrdefinitions.cmake`, so I don't think this has any side effects.

swgillespie commented on clrfeatures.cmake:
```diff
@@ -2,6 +2,12 @@ if(CLR_CMAKE_TARGET_TIZEN_LINUX)
+if(FEATURE_STANDALONE_GC_ONLY)
+  set(FEATURE_EVENT_TRACE 0)
+  set(FEATURE_APPDOMAIN_RESOURCE_MONITORING 0)
+  set(FEATURE_COMINTEROP 0)
```
This is an artifact of earlier iterations of this PR, disabling `FEATURE_COMINTEROP` is not necessary to get the GC to build (and in fact, the VM itself doesn't build without it). I will be removing this in the next commit. Even now, this setting doesn't do anything (the `FEATURE_COMINTEROP` define is defined regardless of these setting of the `FEATURE_COMINTEROP` cmake define.

swgillespie commented on clrfeatures.cmake:
```diff
@@ -2,6 +2,12 @@ if(CLR_CMAKE_TARGET_TIZEN_LINUX)
 
+if(FEATURE_STANDALONE_GC_ONLY)
+  set(FEATURE_EVENT_TRACE 0)
+  set(FEATURE_APPDOMAIN_RESOURCE_MONITORING 0)
```
See https://github.com/dotnet/coreclr/issues/11518

swgillespie commented on src/dlls/mscoree/coreclr/CMakeLists.txt:
```diff
@@ -133,12 +128,6 @@ else()
-    list(APPEND CORECLR_LIBRARIES
-        gc_unix
-    )
-endif(CLR_CMAKE_PLATFORM_UNIX AND FEATURE_STANDALONE_GC)
```
Nothing in this file is necessary unless we are linking a standalone GC into the runtime itself, which we don't need to do anymore because it can be loaded dynamically.

swgillespie commented on src/gc/CMakeLists.txt:
```diff
@@ -26,12 +16,48 @@ set( GC_SOURCES
+endif(CLR_CMAKE_PLATFORM_DARWIN)
+
+target_include_directories(gc BEFORE PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
+target_include_directories(gc BEFORE PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/env)
```
I tried to restrict the compile flags and include directories to this target, to avoid perturbing the rest of the build. It seems to work.

janvorli replied:
I'm not sure what you mean by "avoid perturbing the rest of the build". The options you add using add_compile_options / include_directories take effect only for the current CMakeLists.txt and the CMakeLists.txt in subdirectories it includes.

swgillespie replied:
Ah, seems I was misunderstanding the scope of `add_compile_options` and `include_directories`. Will fix.

swgillespie commented on src/gc/env/gcenv.base.h:
```diff
@@ -7,10 +7,11 @@
+#endif // _MSC_VER
 
-#define GCENV_INCLUDED
+#define FEATURE_REDHAWK 1
```
The standalone GC does build with `FEATURE_REDHAWK` for now, as a stepping stone until the remainder of the Local GC work items are handled.

swgillespie commented on src/gc/env/gcenv.base.h:
```diff
@@ -175,8 +181,130 @@ typedef DWORD (WINAPI *PTHREAD_START_ROUTINE)(void* lpThreadParameter);
+// Only clang defines __has_builtin, so we first test for a GCC define
+// before using __has_builtin.
+
+#if defined(__i386__) || defined(__x86_64__)
```
This section defines compiler intrinsic wrappers and other helpers used in the GC (`ALIGN_UP` and `ALIGN_DOWN`) that are normally defined by the EE.

swgillespie commented on src/gc/env/gcenv.base.h:
```diff
@@ -331,7 +459,7 @@ typedef PTR_PTR_Object PTR_UNCHECKED_OBJECTREF;
 #define OBJECTREFToObject(_obj) (Object*)(_obj)
 
-#define VALIDATEOBJECTREF(_objref) _objref;
+#define VALIDATEOBJECTREF(_objref) (void)_objref;
```
Silences a warning.

swgillespie commented on src/gc/env/gcenv.base.h:
```diff
@@ -508,7 +640,11 @@ void LogSpewAlways(const char *fmt, ...);
+inline bool IsGCThread()
+{
+    // [LOCALGC TODO] this is not correct
+    return false;
```
This (and the above `IsGCSpecialThread` are stubbed out for now (WIP).

swgillespie commented on src/gc/env/gcenv.base.h:
```diff
@@ -508,7 +640,11 @@ void LogSpewAlways(const char *fmt, ...);
+    return false;
+}
 
 class CLRConfig
```
I think this can be removed now that the `GCConfig` class exists.

swgillespie commented on src/gc/env/gcenv.base.h:
```diff
@@ -606,17 +742,47 @@ class GCStress
-    static bool CanEnableGCNumaAware();
-    static void GetGroupForProcessor(uint16_t processor_number, uint16_t * group_number, uint16_t * group_processor_number);
-    static bool GetNumaProcessorNodeEx(PPROCESSOR_NUMBER proc_no, uint16_t * node_no);
+    static bool CanEnableGCNumaAware()
```
This (and `CPUGroupInfo`) are both also stubbed out (WIP).

swgillespie commented on src/gc/env/gcenv.h:
```diff
@@ -0,0 +1,77 @@
+#include "gcenv.sync.h"
+#include "gcinterface.h"
+
+// [LOCALGC TODO] This is kind of a kludge
```
This avoids linker problems. It's been a little while since I wrote this and I don't remember exactly what goes wrong without the anonymous namespace. I'll try and fix it and report back 😄

swgillespie commented on src/gc/env/gcenv.object.h:
```diff
@@ -64,19 +68,31 @@ class MethodTable
+        //
+        // The solution here is to do what the VM does and check the
+        // HasComponentSize flag so that we're on the same page.
+        return (m_flags & MTFlag_HasComponentSize) != 0;
```
This was hard to track down - it should be accurate now with what the VM does with the m_flags field. CoreRT doesn't use this code path (it forwards methods on the MethodTable facade to `EEType`).

swgillespie commented on src/gc/env/gcenv.object.h:
```diff
@@ -33,9 +33,13 @@ class ObjHeader
+#define MTFlag_HasFinalizer     0x0010
+#define MTFlag_IsArray          0x0008
+
+// TODO(segilles) get the correct value for this
```
This is the correct value, I just haven't removed the comment yet...

swgillespie commented on src/gc/env/gcenv.object.h:
```diff
@@ -104,6 +120,12 @@ class MethodTable
+    uint8_t* GetLoaderAllocatorObjectForGC()
+    {
+        // [LOCALGC TODO] this is not correct
+        return nullptr;
```
Another WIP

swgillespie commented on src/gc/objecthandle.cpp:
```diff
@@ -21,12 +21,14 @@
-#endif // FEATURE_REDHAWK
+#ifndef BUILD_AS_STANDALONE
+ #ifdef FEATURE_COMINTEROP
+  #include "comcallablewrapper.h"
```
This header might not be necessary.

swgillespie commented on src/gc/objecthandle.cpp:
```diff
@@ -137,7 +139,7 @@ void CALLBACK TraceDependentHandle(_UNCHECKED_OBJECTREF *pObjRef, uintptr_t *pEx
     // objects are NULL.  However, if the secondary object is non-NULL, then the primary
     // object should also be non-NULL.
-    _ASSERTE(*pExtraInfo == NULL || *pObjRef != NULL);
+    _ASSERTE(*pExtraInfo == 0 || *pObjRef != NULL);
```
Silences a warning.

swgillespie commented on src/gc/objecthandle.h:
```diff
@@ -17,7 +17,7 @@
 #include "handletable.h"
 
-#ifdef FEATURE_COMINTEROP
+#if 0 // [LOCALGC TODO] WinRT weak references
```
Another WIP

swgillespie commented on src/gc/unix/gcenv.unix.cpp:
```diff
@@ -6,28 +6,10 @@
-// clang typedefs uint64_t to be unsigned long long, which clashes with
-// PAL/MSVC's unsigned long, causing linker errors. This ugly hack
-// will go away once the GC doesn't depend on PAL headers.
-typedef unsigned long uint64_t_hack;
```
Very happy to delete this horrible hack I did a few months ago.

swgillespie commented on src/vm/CMakeLists.txt:
```diff
@@ -202,7 +202,6 @@ set(VM_SOURCES_WKS
     gccover.cpp
     gcenv.ee.static.cpp
     gcenv.ee.common.cpp
-    gcenv.os.cpp
```
This was an error - the VM would link in `gcenv.os.cpp` twice if building with a standalone GC and linking it into the runtime, which resulted in linker errors. Removing this source inclusion causes it to build once, as expected.

swgillespie commented on src/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -986,7 +986,7 @@ HRESULT AllowObjectInspection()
 #endif // PROFILING_SUPPORTED
 
-#if defined(GC_PROFILING) || defined(FEATURE_EVENT_TRACE)
+#if defined(PROFILING_SUPPORTED) || defined(FEATURE_EVENT_TRACE)
```
`GC_PROFILING` is a GC-specific define and shouldn't cross the GC/EE interface.

swgillespie commented:
I'm going to fix the `snprintf` compile error and x86 windows issues, but I think this PR is potentially ready for a preliminary review?

jkotas commented on src/gc/env/gcenv.base.h:
```diff
@@ -175,8 +181,130 @@ typedef DWORD (WINAPI *PTHREAD_START_ROUTINE)(void* lpThreadParameter);
+#endif // _MSC_VER
+
+// Cross-platform wrapper for the _BitScanForward compiler intrinsic.
+inline unsigned char BitScanForward(DWORD *bitIndex, DWORD mask)
```
Can this use standard types (uint32_t, etc.) instead?

swgillespie replied:
Sure! I started out by mirroring the type signature expected by uses of this function (`softwarewritewatch.h/cpp`, https://github.com/dotnet/coreclr/blob/master/src/gc/softwarewritewatch.cpp#L78) but I can do standard types.

jkotas commented:
Is the empty `src/gc/WarningControl.h` file intentional?

swgillespie commented:
> Is the empty src/gc/WarningControl.h file intentional?

Yes, it's force-included here: https://github.com/dotnet/coreclr/blob/7f70e548bb22d7e551dd7167863c83e984d0755e/CMakeLists.txt#L574.

swgillespie commented:
The two Windows x64 test failures are both instances of this:

```
Assert failure(PID 19356 [0x00004b9c], Thread: 31636 [0x7b94]): g_fEEShutDown || !"Crst Reentrancy violation"

CORECLR! CrstBase::IsSafeToTake + 0x2C2 (0x00007ffb`893444e2)
CORECLR! CrstBase::Enter + 0x233 (0x00007ffb`89343d93)
CORECLR! ThreadStore::Enter + 0x48 (0x00007ffb`893eb708)
CORECLR! ThreadSuspend::LockThreadStore + 0x2DB (0x00007ffb`893cbe3b)
CORECLR! ThreadSuspend::SuspendEE + 0x11C (0x00007ffb`893d205c)
CORECLR! WriteBarrierManager::ChangeWriteBarrierTo + 0xA0 (0x00007ffb`89a56cf0)
CORECLR! WriteBarrierManager::UpdateWriteWatchAndCardTableLocations + 0x41 (0x00007ffb`89a577b1)
CORECLR! StompWriteBarrierResize + 0x23 (0x00007ffb`89a575e3)
CORECLR! standalone::GCToEEInterface::StompWriteBarrier + 0x110 (0x00007ffb`897165d0)
GC! `anonymous namespace'::GCToEEInterface::StompWriteBarrier + 0x5F (0x00007ffb`b36f56cf)
    File: e:\github\coreclr\src\vm\crst.cpp Line: 719
    Image: E:\GitHub\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\CoreRun.exe



Return code:      1
Raw output file:      E:\GitHub\coreclr\bin\tests\Windows_NT.x64.Debug\Reports\GC.Regressions\v2.0-rtm\494226\494226\494226.output.txt
Raw output:
BEGIN EXECUTION
 "E:\GitHub\coreclr\tests\..\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\corerun.exe" 494226.exe
Expected: 100
Actual: 123456789
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1

To run the test:
> set CORE_ROOT=E:\GitHub\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root
> E:\GitHub\coreclr\bin\tests\Windows_NT.x64.Debug\GC\Regressions\v2.0-rtm\494226\494226\494226.cmd

Expected: True
Actual:   False
```

swgillespie commented:
The last commits fix the build issues, next I'll be fixing the test failures. (Windows x86 standalone GC build is currently broken for reasons unrelated to this PR so I think it can be addressed in a follow-up PR).

swgillespie commented:
I believe I've addressed all issues that should affect the non-standalone GC build with the last commit. I'm still in the process of validating that this is the case but this PR should be ready for reviews with intent to merge.

janvorli commented on src/gc/CMakeLists.txt:
```diff
@@ -26,12 +16,48 @@ set( GC_SOURCES
+
+if(CLR_CMAKE_PLATFORM_UNIX)
+  target_compile_options(gc PUBLIC -fPIC)
+  # dprintf causes many warnings
```
Maybe we should create a GitHub issue to fix that at some point. Lot of the formatting options in dprintf are MSVC specific and fixing that would allow you to debug stuff using those on Unix too.

swgillespie replied:
Done: https://github.com/dotnet/coreclr/issues/13367

janvorli commented on src/gc/env/gcenv.object.h:
```diff
@@ -64,19 +66,31 @@ class MethodTable
         return m_componentSize;
     }
 
+    bool Collectible()
```
A nit - it seems that it would be better to name the function `IsCollectible` (and modify the define accordingly) to match the other flags that are in a form of Verb + Noun.

swgillespie replied:
Agreed - this definition mirrors the one in `methodtable.h`, though: https://github.com/dotnet/coreclr/blob/master/src/vm/methodtable.h#L1912 . We'd need to rename that one too.

janvorli commented on src/gc/objecthandle.cpp:
```diff
@@ -21,12 +21,14 @@
-#ifndef FEATURE_REDHAWK
-#include "nativeoverlapped.h"
-#endif // FEATURE_REDHAWK
+#ifndef BUILD_AS_STANDALONE
```
A nit - could you please revert the indentation of the `#include`s and `#ifdef`s? We don't do that in most of the code.

swgillespie replied:
will do!

janvorli commented on src/gc/unix/configure.cmake:
```diff
@@ -54,4 +54,4 @@ check_cxx_source_runs("
 check_library_exists(c sched_getaffinity "" HAVE_SCHED_GETAFFINITY)
 
-configure_file(${CMAKE_CURRENT_SOURCE_DIR}/config.h.in ${CMAKE_CURRENT_BINARY_DIR}/config.h)
+configure_file(${CMAKE_CURRENT_SOURCE_DIR}/unix/config.h.in ${CMAKE_CURRENT_BINARY_DIR}/config.h)
```
I'd prefer replacing `CMAKE_CURRENT_SOURCE_DIR` by `CMAKE_CURRENT_LIST_DIR` here instead of adding the `/unix/`.

swgillespie replied:
didn't know about `CMAKE_CURRENT_LIST_DIR`, thanks! will do.

swgillespie commented on src/gc/CMakeLists.txt:
```diff
@@ -26,12 +16,48 @@ set( GC_SOURCES
 convert_to_absolute_path(GC_SOURCES ${GC_SOURCES})
 
-add_library_clr(gc_standalone STATIC ${GC_SOURCES})
+add_library(gc SHARED ${GC_SOURCES})
```
I think it will - will do!

swgillespie replied:
So this is a problem because the function `add_library_clr` is defined after `src\gc\CMakeLists.txt` is included. Looking at the body, it looks like it should be OK to move it to `functions.cmake`. I'll give that a try.

swgillespie commented:
cc @sergiy-k (thought I pinged earlier -sorry!)

swgillespie commented:
@janvorli Thanks for the review! Have I addressed your concerns?

swgillespie commented:
thanks for the reviews!
